File: npm/lib/install-planner.js
const fs = require("fs");
const path = require("path");
const { contentRoot } = require("./content-root");
const { fail } = require("./errors");
const { normalize, listFiles } = require("./fs-utils");
const { sha256 } = require("./hash");
const { selectedSkills } = require("./pack-resolver");
const { buildEntry } = require("./entry-builder");

function collectCodingAgentFiles(manifest, selectedCodingAgents, entryMode, selectedPacks) {
  const byRelativePath = new Map();
  const skills = selectedSkills(manifest, selectedPacks);

  for (const codingAgent of selectedCodingAgents) {
    const sourceRoot = path.join(contentRoot, "generated", codingAgent);
    if (!fs.existsSync(sourceRoot)) {
      fail(`Bundled generated CodingAgent not found: ${codingAgent}`);
    }

    for (const file of listFiles(sourceRoot)) {
      const relativePath = normalize(path.relative(sourceRoot, file));
      if (
        manifest.entryPoints[codingAgent] &&
        relativePath === normalize(manifest.entryPoints[codingAgent])
      ) {
        continue;
      }

      const skillName = generatedSkillName(relativePath);
      if (skillName && !skills.has(skillName)) {
        continue;
      }

      addPlannedFile(byRelativePath, relativePath, file, null, fs.readFileSync(file));
    }

    if (entryMode !== "none") {
      addBuiltFile(byRelativePath, buildEntry(manifest, codingAgent, entryMode, selectedPacks));
    }
  }

  for (const packName of selectedPacks) {
    for (const projectFile of manifest.packs[packName].projectFiles || []) {
      const projectFilesRoot = path.join(contentRoot, projectFile.source);
      if (!fs.existsSync(projectFilesRoot)) {
        fail(`Bundled project files not found: ${projectFile.source}`);
      }

      for (const file of listFiles(projectFilesRoot)) {
        const relativePath = normalize(path.join(projectFile.destination, path.relative(projectFilesRoot, file)));
        addPlannedFile(byRelativePath, relativePath, file, null, fs.readFileSync(file));
      }
    }
  }

  return Array.from(byRelativePath.values());
}

function generatedSkillName(relativePath) {
  const parts = normalize(relativePath).split("/");
  const index = parts.indexOf("skills");
  if (index === -1 || index + 1 >= parts.length) {
    return null;
  }

  return parts[index + 1];
}

function addPlannedFile(byRelativePath, relativePath, sourcePath, content, hashSource) {
  addBuiltFile(byRelativePath, {
    relativePath,
    sourcePath,
    content,
    hash: sha256(hashSource)
  });
}

function addBuiltFile(byRelativePath, file) {
  const existing = byRelativePath.get(file.relativePath);
  if (existing) {
    if (existing.hash !== file.hash) {
      fail(`Conflicting bundled files for path: ${file.relativePath}`);
    }

    return;
  }

  byRelativePath.set(file.relativePath, file);
}

module.exports = { collectCodingAgentFiles };

File: tools/idd-tool/Installation/InstallPlanner.cs
internal sealed class InstallPlanner(ContentLayout layout)
{
    public IReadOnlyList<PlannedFile> Collect(
        Manifest manifest,
        IEnumerable<string> codingAgents,
        EntryMode entryMode,
        IReadOnlyList<string> selectedPacks)
    {
        var byRelativePath = new Dictionary<string, PlannedFile>(StringComparer.Ordinal);
        var selectedSkills = ManifestSkillSelector.SelectedSkills(manifest, selectedPacks);

        foreach (var codingAgent in codingAgents)
        {
            var sourceRoot = Path.Combine(layout.GeneratedRoot, codingAgent);
            if (!Directory.Exists(sourceRoot))
            {
                throw new ToolException($"Bundled generated CodingAgent not found: {codingAgent}");
            }

            foreach (var sourcePath in Directory.GetFiles(sourceRoot, "*", SearchOption.AllDirectories))
            {
                var relativePath = PathNormalizer.Normalize(Path.GetRelativePath(sourceRoot, sourcePath));
                if (manifest.EntryPoints.TryGetValue(codingAgent, out var entryPoint) &&
                    StringComparer.Ordinal.Equals(relativePath, PathNormalizer.Normalize(entryPoint)))
                {
                    continue;
                }

                if (TryGetGeneratedSkillName(relativePath, out var skillName) &&
                    !selectedSkills.Contains(skillName))
                {
                    continue;
                }

                AddFile(byRelativePath, relativePath, File.ReadAllBytes(sourcePath));
            }

            if (entryMode != EntryMode.None)
            {
                AddFile(byRelativePath, new EntryBuilder(layout).Build(manifest, codingAgent, entryMode, selectedPacks));
            }
        }

        foreach (var projectFile in selectedPacks.SelectMany(pack => manifest.Packs[pack].ProjectFiles))
        {
            var projectFilesRoot = Path.Combine(layout.ContentRoot, projectFile.Source);
            if (!Directory.Exists(projectFilesRoot))
            {
                throw new ToolException($"Bundled project files not found: {projectFile.Source}");
            }

            foreach (var sourcePath in Directory.GetFiles(projectFilesRoot, "*", SearchOption.AllDirectories))
            {
                var relativePath = PathNormalizer.Normalize(Path.Combine(projectFile.Destination, Path.GetRelativePath(projectFilesRoot, sourcePath)));
                AddFile(byRelativePath, relativePath, File.ReadAllBytes(sourcePath));
            }
        }

        return byRelativePath.Values.ToArray();
    }

    private static bool TryGetGeneratedSkillName(string relativePath, out string skillName)
    {
        var parts = PathNormalizer.Normalize(relativePath).Split('/');
        for (var index = 0; index < parts.Length - 1; index++)
        {
            if (StringComparer.Ordinal.Equals(parts[index], "skills") && index + 1 < parts.Length)
            {
                skillName = parts[index + 1];
                return true;
            }
        }

        skillName = "";
        return false;
    }

    private static void AddFile(Dictionary<string, PlannedFile> byRelativePath, string relativePath, byte[] content) =>
        AddFile(byRelativePath, new PlannedFile(relativePath, content, FileHasher.Sha256(content)));

    private static void AddFile(Dictionary<string, PlannedFile> byRelativePath, PlannedFile plannedFile)
    {
        if (byRelativePath.TryGetValue(plannedFile.RelativePath, out var existing))
        {
            if (!StringComparer.Ordinal.Equals(existing.Hash, plannedFile.Hash))
            {
                throw new ToolException($"Conflicting bundled files for path: {plannedFile.RelativePath}");
            }

            return;
        }

        byRelativePath.Add(plannedFile.RelativePath, plannedFile);
    }
}
